samples: wifi: add Wi-Fi provisioning sample (#8862)
nrfconnect/sdk-nrf: samples/wifi/provisioning/sample_description.rst

.. _wifi_provisioning_sample_desc:

Sample description
##################

.. contents::
   :local:
   :depth: 2

> The Provisioning Service sample demonstrates how to provision a device with Nordic Semiconductor's Wi-Fi chipsets over Bluetooth® Low Energy.

Reviewer: If this is the first instance of it, we should use the full name.

> The Provisioning Service sample demonstrates how to provision a device with Nordic Semiconductor's Wi-Fi chipsets over Bluetooth® Low Energy.

Author: Let's use "Wi-Fi Provisioning Service sample" here.

Requirements
************

The sample supports the following development kit:

.. table-from-sample-yaml::

The sample also requires a smartphone with Nordic Semiconductor's nRF Wi-Fi Provisioner application installed in one of the following versions:

* `nRF Wi-Fi Provisioner mobile app for Android`_
* `nRF Wi-Fi Provisioner mobile app for iOS`_

Overview
********

This sample can be used to provision a Wi-Fi device, that lacks input or output capability.

> The Provisioning Service is divided into three parts:

Author: It refers to this sample.

* Task and event handling component - It handles provisioning related tasks and events.

> * Transport layer - It is based on Bluetooth® Low Energy that exchanges information between the device and the nRF Wi-Fi Provisioner app.

Reviewer 2: ```suggestion
* Transport layer - It is based on Bluetooth Low Energy that exchanges information between the device and the nRF Wi-Fi Provisioner app.
```

* Configuration management component - It manages the provisioning data (in RAM and flash) which is accessed by multiple threads.

Building and running
********************

.. |sample path| replace:: :file:`samples/wifi/provisioning`

.. include:: /includes/build_and_run.txt

Testing
=======

|test_sample|

1. |connect_kit|
#. |connect_terminal|

> #. Observe the following output::
> 
>       wifi_prov: BT Advertising successfully started

Reviewer 2: ```suggestion
#. Observe the following output:

    .. code-block:: console

       wifi_prov: BT Advertising successfully started
```

#. Open nRF Wi-Fi Provisioner application, and enable Bluetooth if it is disabled.

> #. Search and connect to the device you want to provision. If the connection and pairing is established, observe the following output::
> 
>       wifi_prov: BT Connected
>       wifi_prov: BT Pairing Complete

Reviewer 2: ```suggestion
#. Search and connect to the device you want to provision. If the connection and pairing are established, observe the following output:

    .. code-block:: console

       wifi_prov: BT Connected
       wifi_prov: BT Pairing Complete
```

> #. Conduct the access point (AP) scan, enter password if required, and trigger Wi-Fi connection.

Reviewer: OK. @wentong-li, these should be separate steps:

#. Start an access point scan.
#. Enter the password, if required.
#. Establish a Wi-Fi connection.

> #. Conduct the access point (AP) scan, enter password if required, and trigger Wi-Fi connection.

Author: OK

> #. Observe the following output, if the connection is successful::
> 
>       wpa_supp: wlan0: CTRL-EVENT-CONNECTED - Connection to <AP MAC Address> completed [id=0 id_str=]

Reviewer 2: ```suggestion
#. Observe the following output, if the connection is successful:

    .. code-block:: console
    
       wpa_supp: wlan0: CTRL-EVENT-CONNECTED - Connection to <AP MAC Address> completed [id=0 id_str=]
```

* Task and event handling component - It handles provisioning related tasks and events.
* Transport layer - It is based on Bluetooth® Low Energy that exchanges information between the device and the nRF Wi-Fi Provisioner app.
* Configuration management component - It manages the provisioning data (in RAM and flash) which is accessed by multiple threads.

Building and running
********************

.. |sample path| replace:: :file:`samples/wifi/provisioning`

.. include:: /includes/build_and_run.txt

Testing
=======

|test_sample|

1. |connect_kit|
#. |connect_terminal|
#. Observe the following output:

.. code-block:: console

    wifi_prov: BT Advertising successfully started

#. Open nRF Wi-Fi Provisioner application, and enable Bluetooth if it is disabled.
#. Search and connect to the device you want to provision. If the connection and pairing is established, observe the following output:

.. code-block:: console

       wifi_prov: BT Connected
       wifi_prov: BT Pairing Complete

#. Conduct the access point (AP) scan, enter password if required, and trigger Wi-Fi connection.
#. Observe the following output, if the connection is successful:

.. code-block:: console

>        wpa_supp: wlan0: CTRL-EVENT-CONNECTED - Connection to <AP MAC Address> completed [id=0 id_str=]

Reviewer 2: Please check formatting of .. code-block:: console to correct the ordered list.

.. _wifi_provisioning_sample_desc:

Sample description
##################

.. contents::
   :local:
   :depth: 2

The Wi-Fi Provisioning Service sample demonstrates how to provision a device with Nordic Semiconductor's Wi-Fi chipsets over Bluetooth® Low Energy.

Requirements
************

The sample supports the following development kit:

.. table-from-sample-yaml::

The sample requires a smartphone (configurator) with Nordic Semiconductor's nRF Wi-Fi Provisioner app installed in one of the following versions:

* `nRF Wi-Fi Provisioner mobile app for Android`_
* `nRF Wi-Fi Provisioner mobile app for iOS`_

Overview
********

With this sample, you can provision a Wi-Fi device that lacks input or output capability.
The sample is divided into three parts:

> * Task and event handling component - It handles provisioning related tasks and events.

Reviewer: @richabp, I'm not sure on the list formatting, but 'it handles' reads strangely to me. 

A little more natural is 'handles provisioning-related tasks and events.' without a pronoun. Same for other components.

> * Task and event handling component - It handles provisioning related tasks and events.

Reviewer: ```suggestion
* Task and event handling component: Handles provisioning-related tasks and events.
```

> * Transport layer - It is based on Bluetooth Low Energy that exchanges information between the device and the nRF Wi-Fi Provisioner app.

Reviewer: ```suggestion
* Transport layer: Based on Bluetooth Low Energy and exchanges information between the device and the nRF Wi-Fi Provisioner app.
```

> * Configuration management component - It manages the provisioning data (in RAM and flash) which is accessed by multiple threads.

Reviewer: ```suggestion
* Configuration management component: Manages the provisioning data (in RAM and flash) accessed by multiple threads.
```

Building and running
********************

.. |sample path| replace:: :file:`samples/wifi/provisioning`

.. include:: /includes/build_and_run.txt

Testing
=======

|test_sample|

1. |connect_kit|
#. |connect_terminal|
#. Observe the following output:

    .. code-block:: console

       wifi_prov: BT Advertising successfully started

#. Open the nRF Wi-Fi Provisioner app.
#. Enable Bluetooth if it is disabled.
#. Search and connect to the device you want to provision. If the connection and pairing are established, observe the following output:

    .. code-block:: console

       wifi_prov: BT Connected
       wifi_prov: BT Pairing Complete

#. Conduct the access point (AP) scan, enter password if required, and trigger Wi-Fi connection.
#. Observe the following output, if the connection is successful:

    .. code-block:: console

       wpa_supp: wlan0: CTRL-EVENT-CONNECTED - Connection to <AP MAC Address> completed [id=0 id_str=]

Dependencies
************

This sample uses the following `sdk-nrfxlib`_ library:

* :ref:`nrfxlib:nrf_security`

This sample also uses modules that can be found in the following locations
in the |NCS| folder structure:

> * ``modules/lib/nanopb``

Reviewer: ```suggestion
* :file:`modules/lib/nanopb`
```
apply proper filepath formatting to this and header files in lines 88-91, 95.

Author: For the dependency section, the style is copied from other README.rst files. Not sure if we should change it.

In addition, it uses the following Zephyr libraries:

* :ref:`zephyr:bluetooth_api`:

>   * ``include/bluetooth/bluetooth.h``
>   * ``include/bluetooth/conn.h``
>   * ``include/bluetooth/uuid.h``
>   * ``include/bluetooth/gatt.h``

Reviewer: ```suggestion
  * :file:`include/bluetooth/bluetooth.h`
  * :file:`include/bluetooth/conn.h`
  * :file:`include/bluetooth/uuid.h`
  * :file:`include/bluetooth/gatt.h`
```

Changing these based on Pekka's guidance :)

* :ref:`zephyr:settings_api`:

>   * ``include/settings/settings.h``

Reviewer: ```suggestion
  * :file:`include/settings/settings.h`
```

The sample supports the following development kits:

.. table-from-sample-yaml::

The sample also requires an Android or iOS device with nRF Wi-Fi Provisioner
application installed:

#. `Andriod App <https://play.google.com/store/apps/details?id=no.nordicsemi.android.wifi.provisioning>`_
#. `iOS App <https://apps.apple.com/us/app/nrf-wi-fi-provisioner/id1638948698>`_

Overview
********

The goal of Wi-Fi Provisioning Service is to provide a method that allows
user to provision Wi-Fi device, which typically lacks input/output capability.

The service can be largely divided into three parts:

#. Component that handles provisioning-related tasks and events
#. Transport layer used to exchange information between the device and the configurator
#. Component to handle the Provisioning data (in RAM and flash) which  will be accessed
   by multiple threads

Building and running
********************

.. |sample path| replace:: :file:`samples/wifi/provisioning`

If you want to generate .pb.h and .pb.c files when building the project,
CONFIG_WIFI_PROVISIONING_PROTO_BUILD_TIME_GENERATION needs to be y. If you
want to use pre-generated .pb.h and .pb.c files, it needs to be n. This Kconfig

> option is disabled by default.

Author: To avoid possible build error, we decide to remove this kconfig option at this moment. So this part will be removed.

Testing
=======

|test_sample|

1. |connect_kit|
#. |connect_terminal|
#. After initialization, you can see output like::

       wifi_prov: Advertising successfully started

#. Open nRF Wi-Fi Provisioner application, and enbale Bluetooth if it is disabled.
#. Search for and connect to the device you want provision. If connection and pairing
   is done, you can see output like::

       wifi_prov: Connected
       wifi_prov: Pairing Complete

#. After connection and pairing, you can conduct AP scan, input password if required,
   and trigger Wi-Fi connection attempt.
#. If connection succeeds, you can see output like::

       wpa_supp: wlan0: CTRL-EVENT-CONNECTED - Connection
       to <AP MAC Address> completed [id=0 id_str=]

Dependencies
============

This sample uses the following `sdk-nrfxlib`_ library:

* :ref:`nrfxlib:nrf_security`

> This sample also uses modules that can be found in the following locations
> in the |NCS| folder structure:
> 
> * ``modules/lib/nanopb``

Reviewer 3: ```suggestion
This sample also uses modules that can be found in the following locations
in the |NCS| folder structure:

* ``modules/lib/nanopb``
```

Reviewer 3: We are not using nanopb based generation in this PR, so better not mention it anywhere now.

Author: The runtime library is used (encode/decode.h). And CONFIG_NANOPB has to be y. So I think we need to specify here.
